src/file_watcher.rs

```diff
@@ -0,0 +1,1426 @@
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Start the `FileWatcher`
+    ///
+    /// This function forks and has the parent immediately return. The forked
+    /// off child then daemonises to poll for aged-out telemetry files, sending
+    /// them to InfluxDB, and exits once there are no more files to process.
+    pub fn start(&mut self) -> WatcherResult<()> {
+        self.start_with_helpers(&DefaultStartHelpers)
+    }
+
+    fn start_with_helpers(&mut self, helpers: &impl StartHelpers) -> WatcherResult<()> {
+        if var("FUELUP_NO_TELEMETRY").is_ok() {
+            // If telemetry is disabled, immediately return
+            return Ok(());
+        }
+
+        let logfile = &config().map_err(into_recoverable)?.logfile;
+
+        if STARTED.load(Ordering::Relaxed) {
+            return Ok(());
+        } else {
+            STARTED.store(true, Ordering::Relaxed);
+        }
+
+        match helpers.daemonise(logfile) {
+            Ok(Some(pid)) => {
+                // We are the parent, so record the PID then immediately return
+                PID.store(pid.as_raw(), Ordering::Relaxed);
+                return Ok(());
+            }
+            Err(e) => {
+                // Couldn't daemonise, so clear the PID before returning the error
+                STARTED.store(false, Ordering::Relaxed);
+                return Err(e);
+            }
+            Ok(None) => {
+                // We are the child process, so continue as the `FileWatcher`...
+                PID.store(getpid().as_raw(), Ordering::Relaxed);
+            }
+        }
+
+        // From here on, we are no longer the original process, so the caller should
+        // treat errors as fatal. This means that on error the process should exit
+        // immediately as there should not be two identical flows of execution
+        // from the caller's perspective.
+
+        // Cache the client and request so we don't recreate them each time
+        self.client = Some(Client::new());
+        self.request = Some(helpers.build_request(self)?);
+
+        loop {
+            // Enforce a singleton to ensure we are the only process submitting
+            // telemetry to InfluxDB
+            let _lock = helpers.enforce_singleton(&config()?.lockfile)?;
+
+            // Poll for aged-out telemetry files
+            let directory_empty = helpers.poll_directory(self)?;
+
+            if directory_empty {
+                helpers.exit(0);
+            }
+
+            // Sleep for the configured interval before polling again
+            helpers.sleep(config()?.poll_interval);
+        }
+    }
+
+    /// Kill the `FileWatcher` daemon if one is running
+    pub fn kill() -> Result<bool> {
+        let pid = PID.load(Ordering::Relaxed);
+
+        if pid > 0 {
+            kill(Pid::from_raw(pid), SIGKILL)?;
+            PID.store(0, Ordering::Relaxed);
+            return Ok(true);
+        }
+
+        Ok(false)
+    }
```

Comment:
reads a bit cleaner IMO

```suggestion
pub fn kill() -> Result<bool> {
    let pid = PID.load(Ordering::Relaxed);
    if pid <= 0 {
        return Ok(false);
    }
    
    kill(Pid::from_raw(pid), SIGKILL)?;
    PID.store(0, Ordering::Relaxed);
    Ok(true)
}
```

Reply:
Nice. Although I've changed it to be when `pid == 0` given than anything other than `0` means something - 4a8fe2a.

Reply:
This seems to be addressed, feel free to re-open if you think otherwise: resolving the conversation.